internal/pkg/agent/application/application.go

```diff
@@ -111,10 +116,18 @@ func New(
 	}
 	monitor := monitoring.New(isMonitoringSupported, cfg.Settings.DownloadConfig.OS(), cfg.Settings.MonitoringConfig, agentInfo)
 
+	controlAddress := control.Address()
+	log.Debugf("Control address: %s", controlAddress)
+	address, err := deriveCommsAddress(controlAddress, cfg.Settings.GRPC)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to derive comms GRPC: %w", err)
+	}
+	log.Debugf("Comms GRPS address: %s", address)
+
 	runtime, err := runtime.NewManager(
 		log,
 		baseLogger,
-		cfg.Settings.GRPC.String(),
+		address,
```

Comment:
I think it would be better to just pass the whole `cfg.Settings.GRPC` into the `runtime.NewManager` and allow the manager to create the correct address to communicate. Don't like it outside of the manager, because then different callers could come to different conclusions.